<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
    <head>
        <title>Software development : EAGLE - Simulating Reality - from reference to sample genome</title>
        <link rel="stylesheet" href="styles/site.css" type="text/css" />
        <META http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>

    <body>
        <div id="page">
            <div id="main">
                <div id="main-header" class="pageSectionHeader">
                    <h1 id="title-heading" class="pagetitle">
                                                <span id="title-text">
                            Software development : EAGLE - Simulating Reality - from reference to sample genome
                        </span>
                    </h1>

                    <div class="page-metadata">
                        <p>This page last changed on Sep 04, 2012 by <font color="#0050B2">ljanin</font>.</p>
                    </div>
                </div>

                <div id="content" class="view">
                    <div id="main-content" class="wiki-content group">
                    <h1 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Introduction">Introduction</h1><p>Some of our experts' observations (see <a href="EAGLE%2B-%2BSimulating%2BReality%2B-%2BStrategy%2B-%2Bobservations-metrics-simulation.html">EAGLE - Simulating Reality - Strategy - observations-metrics-simulation</a>) may be due to an improper sample genome, rather than bad sequencing simulation.</p><p>The reference genome contains many 'N' regions.<br /> A real sample genome wouldn't contain any.<br /> What should our Genome Mutator tool do about this?</p><h1 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-The&#39;N&#39;sintheHumanreferencegenome">The 'N's in the Human reference genome</h1><p>From the UCSC gaps file:</p><div class="table-wrap"><table class="confluenceTable"><tbody><tr><th class="confluenceTh"><p>gap type</p></th><th class="confluenceTh"><p>#N</p></th><th class="confluenceTh"><p>%ref (3,095,693,983 bases)</p></th></tr><tr><td class="confluenceTd"><p>centromere</p></td><td class="confluenceTd"><p>72000000</p></td><td class="confluenceTd"><p>2.33%</p></td></tr><tr><td class="confluenceTd"><p>telomere</p></td><td class="confluenceTd"><p>460000</p></td><td class="confluenceTd"><p>0.01%</p></td></tr><tr><td class="confluenceTd"><p>short_arm</p></td><td class="confluenceTd"><p>67161193</p></td><td class="confluenceTd"><p>2.17%</p></td></tr><tr><td class="confluenceTd"><p>heterochromatin</p></td><td class="confluenceTd"><p>72477000</p></td><td class="confluenceTd"><p>2.34%</p></td></tr><tr><td class="confluenceTd"><p>clone</p></td><td class="confluenceTd"><p>6126623</p></td><td class="confluenceTd"><p>0.20%</p></td></tr><tr><td class="confluenceTd"><p>contig</p></td><td class="confluenceTd"><p>16119990</p></td><td class="confluenceTd"><p>0.52%</p></td></tr><tr><td class="confluenceTd"><p>all</p></td><td class="confluenceTd"><p>234344806</p></td><td class="confluenceTd"><p>7.57%</p></td></tr></tbody></table></div><p>Where do they come from?<br /> What are we supposed to replace them with?</p><p>What we found so far:</p><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Repeatsingeneral">Repeats in general</h3><p>From <a href="http://www.sciencemag.org/content/291/5507/1304.full" class="external-link" rel="nofollow">The Sequence of the Human Genome</a>(2001):<br /> <img class="confluence-embedded-image" src="attachments/1712205/1638863.jpg" data-image-src="attachments/1712205/1638863.jpg"></p><p>From <a href="http://www.biomedcentral.com/1471-2105/7/541" class="external-link" rel="nofollow">How repetitive are genomes?</a>(2006):</p><blockquote><p>&quot;Since the sequencing of complex genomes these observations have been made precise: approximately 50% of the human genome is made up of repetitive sequences [<a href="http://www.biomedcentral.com/1471-2105/7/541#B8" class="external-link" rel="nofollow">8</a>]. However, the term &quot;repetitive sequences&quot; encompasses a rather heterogeneous set of elements: 45% of the human genome is covered by transposons, 3% are repeats of less than a hundred base pairs (microsatellites and minisatellites), and 5% consist of recent duplications of large segments of DNA.&quot;</p></blockquote><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-ALUrepeats">ALU repeats</h3><p><a href="http://genome.cshlp.org/content/21/6/840.full" class="external-link" rel="nofollow">Alu repeat discovery and characterization within human genomes</a></p><blockquote><p>&quot;We discover 4342 Alu insertions not found in the human reference genome and show that 98% of a selected subset (63/64) experimentally validate. Of these new insertions, 89% correspond to AluY elements, suggesting that they arose by retrotransposition.&quot;</p></blockquote><p>More than one million Alu retrotransposons comprise over 10% of the human genome sequence.</p><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Telomeres">Telomeres</h3><p><a href="http://en.wikipedia.org/wiki/Telomere" class="external-link" rel="nofollow">http://en.wikipedia.org/wiki/Telomere</a></p><p>Telomeres are supposed to be made of the repeated nucleotide sequence TTAGGG. This &quot;TERC&quot; varies between 3 and 20 kilobases in length.</p><p>In the reference, each telomere is using a 10k window at the beginning and end of each chromosome.<br /> Except chromosome 17, which doesn't have any telomere!?</p><p>The usual pile-up at chr1:10,000 shows the importance of this CCCTAA repeat (reverse-complemented TTAGGG) with a few indel errors.</p><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Subtelomericrepeats">Subtelomeric repeats</h3><p>100-300 kilobases of telomere-associated repeats are between the telomere and the rest of the chromosome.<br /> They contain motifs similar to TTAGGG but, unlike telomeric repeats, they are not conserved in vertebrates.</p><p><a href="http://genome.cshlp.org/content/14/1/18.long" class="external-link" rel="nofollow">Mapping and Initial Analysis of Human Subtelomeric Sequence Assemblies</a>: Figure 2 is great.</p><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Centromeres">Centromeres</h3><h4 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Alphasatellitesequences">Alpha satellite sequences</h4><p><a href="http://www.sciencemag.org/content/291/5507/1304.full" class="external-link" rel="nofollow">The Sequence of the Human Genome</a></p><blockquote><p>&quot;About 5% of the sequence reads were identified as alpha satellite sequences; these were not included in the assembly.&quot;</p></blockquote><p>It is the only repetitive motif associated with centromere function</p><p><a href="http://jhc.sagepub.com/content/54/12/1363.full" class="external-link" rel="nofollow">Chromosome-specific DNA Repeat Probes</a></p><blockquote><p>Generally, the basic repeat units of α-satellite DNA are divergent AT-rich monomers of ~171 bp organized in chromosome-specific higher-order repeat units.<br /> α-Satellite DNA clusters most often contain monomer variants that differ from the consensus sequence by up to 40%.</p><p>At least 33 different alphoid subfamilies have been identified to date based on their repeat organization.</p></blockquote><p>Best one: <a href="http://genome.cshlp.org/content/16/1/88.full" class="external-link" rel="nofollow">The evolutionary dynamics of α-satellite</a></p><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Pericentromericregions">Pericentromeric regions</h3><p>Chr16: <a href="http://hmg.oxfordjournals.org/content/9/1/113.full" class="external-link" rel="nofollow">Molecular structure and evolution of an alpha satellite/non-alpha satellite junction at 16p11</a></p><p>Chr17: <a href="http://mcb.asm.org/content/6/9/3156" class="external-link" rel="nofollow">Structure, organization, and sequence of alpha satellite DNA from human chromosome 17...</a></p><blockquote><p>&quot;On human chromosome 17, the predominant form of alpha satellite is a 2.7-kilobase-pair higher-order repeat unit consisting of 16 alphoid monomers. We present the complete nucleotide sequence of the 16-monomer repeat, which is present in 500 to 1,000 copies per chromosome 17, as well as that of a less abundant 15-monomer repeat, also from chromosome 17.&quot;</p></blockquote><h1 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Other">Other</h1><h3 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-DbSNP">DbSNP</h3><p>Those are not about the Ns, but we could use dbSNP as a source for random generation of SNPs.</p><h1 id="EAGLE-SimulatingReality-fromreferencetosamplegenome-Conclusions">Conclusions</h1>
                    </div>

                                        <div class="pageSection group">
                        <div class="pageSectionHeader">
                            <h2 id="attachments" class="pageSectionTitle">Attachments:</h2>
                        </div>

                        <div class="greybox" align="left">
                                                            <img src="images/icons/bullet_blue.gif" height="8" width="8" alt=""/>
                                <a href="attachments/1712205/1638863.jpg">table 14.jpg</a> (image/jpeg)
                                <br/>
                                                    </div>
                    </div>
                    
                 
                </div>             </div> 
            <div id="footer" style="background: url(http://ukch-confluence.illumina.com/images/border/border_bottom.gif) repeat-x;">
                <p><small>Document generated by Confluence on Jan 29, 2014 12:01</small></p>
            </div>
        </div>     </body>
</html>
