<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
    <head>
        <title>Software development : EAGLE Genome Mutator - Architecture and Design</title>
        <link rel="stylesheet" href="styles/site.css" type="text/css" />
        <META http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>

    <body>
        <div id="page">
            <div id="main">
                <div id="main-header" class="pageSectionHeader">
                    <h1 id="title-heading" class="pagetitle">
                                                <span id="title-text">
                            Software development : EAGLE Genome Mutator - Architecture and Design
                        </span>
                    </h1>

                    <div class="page-metadata">
                        <p>This page last changed on Nov 02, 2011 by <font color="#0050B2">ljanin</font>.</p>
                    </div>
                </div>

                <div id="content" class="view">
                    <div id="main-content" class="wiki-content group">
                    <h1 id="EAGLEGenomeMutator-ArchitectureandDesign-Introduction">Introduction</h1>

<p>The genome mutator (previously called <em>variant modeller</em>) went through a series of architectures before reaching the current one.<br />
In order to avoid repeating the same mistakes, each architecture/design is quickly described here.</p>

<h1 id="EAGLEGenomeMutator-ArchitectureandDesign-Architecture1:Firstprototype">Architecture 1: First prototype</h1>

<p>The list of variants is read to RAM as a double-linked list.<br />
It is then sorted (by variant starting position), and a double-linked list of segments is produced.<br />
The list of segments is then re-ordered by taking into account translocation and deletion variants, which skip some variants.<br />
The output is generated by following the list of segments, fetching the appropriate contigs and writing them to a new fasta file.</p>

<p><img class="confluence-embedded-image" height="242" width="324" src="attachments/1709470/1638720.jpg" data-image-src="attachments/1709470/1638720.jpg"></p>

<p>Problems:</p>
<ul>
	<li>Processing time becomes too high when millions of variants need to be handled.</li>
	<li>Processing reverse strands had some cases not implemented.</li>
</ul>


<h1 id="EAGLEGenomeMutator-ArchitectureandDesign-Architecture2:taggedversion0.1">Architecture 2: tagged version 0.1</h1>

<p>The list of variants is read to RAM as a deque and then sorted.<br />
All the input reference fasta files are also read to RAM.<br />
For each chromosome, its set of alleles (e.g. 2 for a diploid chromosome output) is processed together in order to be able to handle conflicts between variants: when 2 variants are conflicting (modifying the same reference base), one allele would follow one path of variants while the other allele would follow the other path.</p>

<p>The main architecture aspect was the sublist of variants dynamically created for each allele for each contig to be processed. It was dynamically updated if a translocation was making it jump to another chromosome, and wasn't including events that were in the opposite direction than the one being currently processed.</p>

<p>A complex set of classes was introduced for this (right-click to see more details):<br />
<img class="confluence-embedded-image" height="364" width="1009" src="attachments/1709470/1638717.png" data-image-src="attachments/1709470/1638717.png"></p>

<p>Problems:</p>
<ul>
	<li>Small bugs (especially to handle reverse strand processing) kept appearing.</li>
	<li>Handling of the detected conflicts became too complex: the necessity to count how many times each side of a translocation was used became apparent and we decided to move to the next architecture</li>
</ul>


<h1 id="EAGLEGenomeMutator-ArchitectureandDesign-Architecture3:currentversion">Architecture 3: current version</h1>

<p>This is a simplified version of the previous architecture, where the variants list is kept as a vector and an iterator is used to walk through the variants.<br />
Each variant is paired to its &quot;opposite&quot; variant, which is a separate object for translocations and deletions, but is itself in the case of SNP and insertions (as they appear at the same position regardless of the current processed strand direction).<br />
The iterator skips the already-processed variants and those representing variants in the opposite direction, applies the first valid variant and jumps to the paired variant to continue its processing. Marking both variants as processed allows us to process each variant only once. Conflicting variants are automatically skipped when the first allele is being processed (because of the jump to the paired event), and will be processed later on subsequent alleles. If there are more conflicts than alleles, some variants are detected and reported as &quot;not-applied&quot; at the end of the run.</p>

<p><img class="confluence-embedded-image" height="496" width="992" src="attachments/1709470/1638719.png" data-image-src="attachments/1709470/1638719.png"></p>
                    </div>

                                        <div class="pageSection group">
                        <div class="pageSectionHeader">
                            <h2 id="attachments" class="pageSectionTitle">Attachments:</h2>
                        </div>

                        <div class="greybox" align="left">
                                                            <img src="images/icons/bullet_blue.gif" height="8" width="8" alt=""/>
                                <a href="attachments/1709470/1638716.dot">EagleClassDiagram1.dot</a> (application/msword)
                                <br/>
                                                            <img src="images/icons/bullet_blue.gif" height="8" width="8" alt=""/>
                                <a href="attachments/1709470/1638717.png">EagleClassDiagram1.png</a> (image/png)
                                <br/>
                                                            <img src="images/icons/bullet_blue.gif" height="8" width="8" alt=""/>
                                <a href="attachments/1709470/1638718.dot">EagleClassDiagram2.dot</a> (application/msword)
                                <br/>
                                                            <img src="images/icons/bullet_blue.gif" height="8" width="8" alt=""/>
                                <a href="attachments/1709470/1638719.png">EagleClassDiagram2.png</a> (image/png)
                                <br/>
                                                            <img src="images/icons/bullet_blue.gif" height="8" width="8" alt=""/>
                                <a href="attachments/1709470/1638720.jpg">Eagle prototype1 draft.jpg</a> (image/jpeg)
                                <br/>
                                                    </div>
                    </div>
                    
                 
                </div>             </div> 
            <div id="footer" style="background: url(http://ukch-confluence.illumina.com/images/border/border_bottom.gif) repeat-x;">
                <p><small>Document generated by Confluence on Jan 29, 2014 12:01</small></p>
            </div>
        </div>     </body>
</html>
