<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
    <head>
        <title>Software development : EAGLE - For Nondas - On simulating 2x250 long reads</title>
        <link rel="stylesheet" href="styles/site.css" type="text/css" />
        <META http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>

    <body>
        <div id="page">
            <div id="main">
                <div id="main-header" class="pageSectionHeader">
                    <h1 id="title-heading" class="pagetitle">
                                                <span id="title-text">
                            Software development : EAGLE - For Nondas - On simulating 2x250 long reads
                        </span>
                    </h1>

                    <div class="page-metadata">
                        <p>This page last changed on Apr 29, 2013 by <font color="#0050B2">ljanin</font>.</p>
                    </div>
                </div>

                <div id="content" class="view">
                    <div id="main-content" class="wiki-content group">
                    <h1 id="EAGLE-ForNondas-Onsimulating2x250longreads-RunningEAGLE">Running EAGLE</h1><p>Note: I did all this with &quot;ssh ukch-tst-lnts08&quot;</p><p>Note 2: We are actually simulating 2x251 cycles</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-Chr20only">Chr20 only</h3><p>source /illumina/development/EAGLE/EAGLE-snapshot-20130429-setup</p><p>configureEAGLE.pl \<br />  --reference-genome=/illumina/scratch/eagle/References/iGenomes_hg19_with_fai/chr20.fa \<br />  --variant-list=/illumina/scratch/eagle/RealisticInputs/dbSnp_set2_UCSC_chr20.vcf \<br />  --run-info=$EAGLE_SHARE/RunInfo/RunInfo_PairedReads2x251Cycles2x64Tiles.xml \<br />  --template-length-table=$EAGLE_SHARE/TemplateLengthTables/TemplateLengthTableFrom2x250Run.tsv \<br />  --quality-table=${EAGLE_SHARE}/QualityTables/NewQualityTable.read1.length251.qtable2 \<br />  --quality-table=${EAGLE_SHARE}/QualityTables/NewQualityTable.read2.length251.qtable2 \<br />  --coverage-depth=40<br /><br /> </p><p>make -C EAGLE -j 24</p><p>Note: Maybe try with an extra --variant-list=/illumina/scratch/eagle/RealisticInputs/homopolymerIndels_UCSC_chr20.vcf \</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-Fullgenome">Full genome</h3><p>source /illumina/development/EAGLE/EAGLE-1.5-setup</p><p>configureEAGLE.pl \<br />  --reference-genome=/illumina/scratch/eagle/References/iGenomes_hg19_single_file/genome.fa \<br />  --variant-list=/illumina/scratch/eagle/RealisticInputs/dbSnp_set2_UCSC.vcf \<br />  --run-info=$EAGLE_SHARE/RunInfo/RunInfo_PairedReads2x251Cycles2x64Tiles.xml \<br />  --template-length-table=$EAGLE_SHARE/TemplateLengthTables/TemplateLengthTableFrom2x250Run.tsv \<br />  --quality-table=${EAGLE_SHARE}/QualityTables/NewQualityTable.read1.length251.qtable2 \<br />  --quality-table=${EAGLE_SHARE}/QualityTables/NewQualityTable.read2.length251.qtable2 \<br />  --coverage-depth=40</p><p>make -C EAGLE -j 24</p><p>Note: Maybe try with an extra --variant-list=/illumina/scratch/eagle/RealisticInputs/homopolymerIndels_UCSC.vcf \</p><h1 id="EAGLE-ForNondas-Onsimulating2x250longreads-RunningiSAAC">Running iSAAC</h1><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-Chr20only.1">Chr20 only</h3><p>/illumina/development/iSAAC/stable/bin/isaac-align \<br />  -b EAGLE/121221_HSQ868_0359_AH04ADADXX/Data/Intensities/BaseCalls \<br />  -r /illumina/scratch/eagle/References/forISAAC/HumanChr20/sorted-reference.xml \<br />  -m 80</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-Fullgenome.1">Full genome</h3><p>/illumina/development/iSAAC/stable/bin/isaac-align \<br />  -b EAGLE/121221_HSQ868_0359_AH04ADADXX/Data/Intensities/BaseCalls \<br />  -r /illumina/development/iSAAC/iGenomes/Homo_sapiens/UCSC/hg19/Sequence/iSAACIndex.20121213/sorted-reference.xml \<br />  -m 80</p><h1 id="EAGLE-ForNondas-Onsimulating2x250longreads-CustomisingEAGLE">Customising EAGLE</h1><p>You might be interested in customising the following inputs:</p><ul><li>QualityTable</li><li>HomopolymerQualityDropTable</li><li>TemplateLengthTable</li><li>QQTable</li></ul><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-QualityTable">QualityTable</h3><p>You need 1 quality table per read, and pass them to configureEAGLE.pl using the following style:</p><p>configureEAGLE.pl \<br />  --quality-table=qualityTable.length251.read1.qval \<br />  --quality-table=qualityTable.length251.read2.qval \<br />  ...</p><p>Generating a new quality distribution table is easy if you have an existing Run Folder, which you would like to &quot;clone&quot;. (For this, see below the section where I describe how I generated the current quality table).</p><p>Generating a new file by hand may be painful. The format is a tsv file with the following columns:</p><ul style="list-style-type: square;"><li>column 1: cycle number, starting from 1 for each read (&quot;C&quot;)</li><li>column 2: profile number (&quot;P&quot;)</li><li>columns 3-43 (&quot;Q0&quot;, &quot;Q1&quot;, ..., &quot;Q40&quot;): probability for profile &quot;P&quot; to give us a quality &quot;Qx&quot; at cycle &quot;C&quot;<ul style="list-style-type: square;"><li>There is 1 special case when &quot;C&quot;=1 and &quot;Q&quot;=0, which is used to describe the probability for each profile</li></ul></li></ul><p>An easy file to start from is $EAGLE_SHARE/QualityTables/Q10QualityTable.length101.qval</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-HomopolymerQualityDropTable">HomopolymerQualityDropTable</h3><p>I placed 3 tables in $EAGLE_SHARE/HomopolymerQualityDropTables:</p><ul style="list-style-type: square;"><li>the default table: DefaultHomopolymerQualityDropTable.tsv</li><li>a relaxed version of this table: HomopolymerQualityDropTableRelaxed50pc.tsv</li><li>a less relaxed version of this table: HomopolymerQualityDropTableRelaxed33pc.tsv</li><li>a table with no quality drop at all: ZeroHomopolymerQualityDropTable.tsv</li></ul><p>They can be used by passing an extra parameter to configureEAGLE.pl: --homopolymer-quality-drop-table=&lt;filename&gt;</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-TemplateLengthTable">TemplateLengthTable</h3><p>Tab-separated file with 2 columns: template length and count (or probability of occurrence)</p><p>Example: $EAGLE_SHARE/TemplateLengthTables/TemplateLengthTableFrom2x250Run.tsv</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-QQTable">QQTable</h3><p>Tab-separated file with 2 columns: template length and count (or probability of occurrence)</p><p>Example: $EAGLE_SHARE/QualityTables/DefaultQQTable.tsv</p><h1 id="EAGLE-ForNondas-Onsimulating2x250longreads-HowIgeneratedthecurrentinputtables">How I generated the current input tables</h1><ul><li>RunInfo</li><li>TemplateLengthTable</li><li>HomopolymerQualityDropTable</li><li>QualityTable</li></ul><p><em>Default tables were used for:</em></p><ul><li><em>HomopolymerIndelTable</em></li><li><em>MismatchTable</em></li><li><em>QQTable</em></li></ul><h2 id="EAGLE-ForNondas-Onsimulating2x250longreads-RunInfo">RunInfo</h2><p>cp /illumina/scratch/chimera/Chimera_JPM/UDT_173_XL/primary/121221_HSQ868_0359_AH04ADADXX/RunInfo.xml RunInfo/</p><h2 id="EAGLE-ForNondas-Onsimulating2x250longreads-TemplateLengthTable.1">TemplateLengthTable</h2><p>Extraction from real run's pam metrics:</p><p>for i in x y; do ( grep -A 16 InsertSizeHist /illumina/builds/chimera/Chimera_JPM/UDT_173_XL.isis/PUMA.PAM.21Jan/pam/PairStats.txt | grep -A 3 &quot;\&quot;$i\&quot;&quot; | grep values | sed 's/, /\n/g' |tail -n +250 |head -1000 ) &gt; axis_$i ; done</p><p>paste axis_x axis_y &gt; templateLengthTable.tsv</p><h3 id="EAGLE-ForNondas-Onsimulating2x250longreads-HomopolymerQualityDropTable.1">HomopolymerQualityDropTable</h3><p>Extracted from real run, now available as default table</p><p>cd /illumina/scratch/chimera/Chimera_JPM/UDT_173_XL/primary/121221_HSQ868_0359_AH04ADADXX/Data/Intensities/BaseCalls/L001</p><p>$EAGLE_LIBEXEC/generateHomopolymerQualityDropTable.pl -l . -o DefaultHomopolymerQualityDropTable.tsv</p><h4 id="EAGLE-ForNondas-Onsimulating2x250longreads-Tablerelaxed50%">Table relaxed 50%</h4><p>grep -vP '^#' /home/ljanin/install/EAGLE-release/share/EAGLE/HomopolymerQualityDropTables/DefaultHomopolymerQualityDropTable.tsv | awk 'BEGIN { FS=OFS=&quot;\t&quot; } { for (i=2; i&lt;=NF; i++) { $i = ($i + 100)/2 } ; print }'  &gt; relaxed50pcHomopolymerQualityDropTable.tsv</p><h4 id="EAGLE-ForNondas-Onsimulating2x250longreads-Tablerelaxed33%">Table relaxed 33%</h4><p>grep -vP '^#' /home/ljanin/install/EAGLE-release/share/EAGLE/HomopolymerQualityDropTables/DefaultHomopolymerQualityDropTable.tsv | awk 'BEGIN { FS=OFS=&quot;\t&quot; } { for (i=2; i&lt;=NF; i++) { $i = ($i * 2 + 100)/3 } ; print }'  &gt; relaxed33pcHomopolymerQualityDropTable.tsv</p><h2 id="EAGLE-ForNondas-Onsimulating2x250longreads-QualityTable.1">QualityTable</h2><h4 id="EAGLE-ForNondas-Onsimulating2x250longreads-Generationofanintermediatefile">Generation of an intermediate file</h4><p>time ~/workspace/git/EAGLE/src/perl/libexec/generateQualityTable.pl -l /illumina/scratch/chimera/Chimera_JPM/UDT_173_XL/primary/121221_HSQ868_0359_AH04ADADXX/Data/Intensities/BaseCalls/L001/ -t s_1_1101 -o out_s_1_1101.qval</p><h4 id="EAGLE-ForNondas-Onsimulating2x250longreads-Splitofthefileinto2readswithasharedheader">Split of the file into 2 reads with a shared header</h4><p>head -2 out_s_1_1101.qval.format2 | tail -n +2 &gt; header</p><p>(cat header ; awk 'BEGIN { OFS=&quot;\t&quot; } { if (NF == 43 &amp;&amp; $1 &lt;= 251 &amp;&amp; $2&gt;0) {          print $0 } }' out_s_1_1101.qval.format2 ) &gt; qualityTable0.length251.read1.qval</p><p>(cat header ; awk 'BEGIN { OFS=&quot;\t&quot; } { if (NF == 43 &amp;&amp; $1 &gt;  251 &amp;&amp; $2&gt;0) { $1-=251; print $0 } }' out_s_1_1101.qval.format2 ) &gt; qualityTable0.length251.read2.qval</p><h2 id="EAGLE-ForNondas-Onsimulating2x250longreads-NewQualityTable">New Quality Table</h2><p>time /home/ljanin/workspace/build/EAGLE-release/c++/bin/eagleAnalysis_bclAnalyser -r /illumina/upload/std/cleanup/121221_HSQ868_0359_AH04ADADXX --first-cycle=1 --last-cycle=251 -o outRead1 &gt; outRead1.out</p><p>time /home/ljanin/workspace/build/EAGLE-release/c++/bin/eagleAnalysis_bclAnalyser -r /illumina/upload/std/cleanup/121221_HSQ868_0359_AH04ADADXX --first-cycle=252 --last-cycle=502 -o outRead2 &gt; outRead2.out</p><p>=&gt; The generated files outRead1.qtable2 and outRead2.qtable2 can be used as quality tables in the simulator.</p><p>Then we can generate the motifQualityDropTable:</p><p>sort -S 150G -k1,2 outRead1.out &gt; outRead1.out.sorted</p><p>cut -f 1,2,3,4 outRead1.out.sorted | awk 'BEGIN { c1=0; c2=0; c3=0; c4=0; countC4=0; dist=&quot;&quot;; tot=0; count=1 } { if (c4==$4) { countC4+=1 } else { dist=dist &quot;\t&quot; c4 &quot;:&quot; countC4 ; countC4=1; c4=$4 }; if (c1==$1 &amp;&amp; c2==$2) { } else { if (count&gt;10) { print c1 &quot;\t&quot; c2 &quot;\t&quot; c3 &quot;\t&quot; (tot/count) dist } ; c1=$1 ; c2=$2 ; c3=$3 ; c4=$4; countC4=1; dist=&quot;&quot;; tot=0; count=0 } ; tot+=$4 ; count+=1 }' &gt; qualityDropTable</p><p>=&gt; qualityDropTable can be used in the simulator</p><p> </p>
                    </div>

                    
                 
                </div>             </div> 
            <div id="footer" style="background: url(http://ukch-confluence.illumina.com/images/border/border_bottom.gif) repeat-x;">
                <p><small>Document generated by Confluence on Jan 29, 2014 12:01</small></p>
            </div>
        </div>     </body>
</html>
